File: frontend/src/utils/reddit.js
const HEADERS = { Accept: 'application/json' }
const BASE = 'https://www.reddit.com'

export async function fetchByKeyword(subreddit, keyword, limit = 25) {
  return _parse(`${BASE}/r/${subreddit}/search.json?q=${encodeURIComponent(keyword)}&restrict_sr=1&limit=${limit}&sort=new`)
}

export async function fetchRecentPosts(subreddit, limit = 25) {
  return _parse(`${BASE}/r/${subreddit}/new.json?limit=${limit}`)
}

export async function fetchUserPosts(username, limit = 25) {
  return _parse(`${BASE}/user/${username}/submitted.json?limit=${limit}&sort=new`)
}

export async function fetchUserComments(username, limit = 25) {
  const res = await fetch(`${BASE}/user/${username}/comments.json?limit=${limit}`, { headers: HEADERS })
  if (!res.ok) throw new Error(`Reddit error: ${res.status}`)
  const data = await res.json()
  return data.data.children.map(({ data: c }) => ({
    id: c.id,
    title: `Comment in r/${c.subreddit}`,
    text: c.body,
    upvotes: c.score,
    author: c.author,
    created: c.created_utc,
    numComments: 0,
    url: `https://reddit.com${c.permalink}`,
  }))
}

async function _parse(url) {
  const res = await fetch(url, { headers: HEADERS })
  if (!res.ok) throw new Error(`Reddit error: ${res.status}`)
  const data = await res.json()
  return data.data.children.map(({ data: p }) => ({
    id: p.id,
    title: p.title,
    text: p.selftext,
    upvotes: p.score,
    author: p.author,
    created: p.created_utc,
    numComments: p.num_comments,
    url: `https://reddit.com${p.permalink}`,
  }))
}

File: backend/reddit.py
import requests

HEADERS = {"User-Agent": "reddit-sentiment-app/1.0"}
BASE = "https://www.reddit.com"


def fetch_by_keyword(subreddit: str, keyword: str, limit: int = 25):
    url = f"{BASE}/r/{subreddit}/search.json?q={keyword}&restrict_sr=1&limit={limit}&sort=new"
    return _parse(url)


def fetch_recent(subreddit: str, limit: int = 25):
    return _parse(f"{BASE}/r/{subreddit}/new.json?limit={limit}")


def fetch_user_posts(username: str, limit: int = 25):
    return _parse(f"{BASE}/user/{username}/submitted.json?limit={limit}&sort=new")


def fetch_user_comments(username: str, limit: int = 25):
    r = requests.get(
        f"{BASE}/user/{username}/comments.json?limit={limit}",
        headers=HEADERS,
    )
    r.raise_for_status()
    return [
        {
            "id": c["data"]["id"],
            "title": f"Comment in r/{c['data']['subreddit']}",
            "text": c["data"]["body"],
            "upvotes": c["data"]["score"],
            "author": c["data"]["author"],
            "created": c["data"]["created_utc"],
            "numComments": 0,
            "url": f"https://reddit.com{c['data']['permalink']}",
        }
        for c in r.json()["data"]["children"]
    ]


def _parse(url: str):
    r = requests.get(url, headers=HEADERS)
    r.raise_for_status()
    return [
        {
            "id": p["data"]["id"],
            "title": p["data"]["title"],
            "text": p["data"]["selftext"],
            "upvotes": p["data"]["score"],
            "author": p["data"]["author"],
            "created": p["data"]["created_utc"],
            "numComments": p["data"]["num_comments"],
            "url": f"https://reddit.com{p['data']['permalink']}",
        }
        for p in r.json()["data"]["children"]
    ]